arcgis-ios-sdk-samples/Route and directions/Navigate route with rerouting/NavigateRouteWithReroutingViewController.swift

```diff
@@ -0,0 +1,364 @@
+            self.presentAlert(error: error)
+        } else {
+            // Get the new directions.
+            directionsList = (trackingStatus?.routeResult.routes.first!.directionManeuvers)!
+        }
+    }
+    
+    func setSpeakDirection(with text: String) {
+        speechSynthesizer.stopSpeaking(at: .word)
+        speechSynthesizer.speak(AVSpeechUtterance(string: text))
+    }
+    
+    func updateTrackingStatusDisplay(routeTracker: AGSRouteTracker, status: AGSTrackingStatus) {
+        var statusText: String
+        switch status.destinationStatus {
+        case .notReached, .approaching:
+            let distanceRemaining = status.routeProgress.remainingDistance.displayText + " " + status.routeProgress.remainingDistance.displayTextUnits.abbreviation
+            let timeRemaining = timeFormatter.string(from: TimeInterval(status.routeProgress.remainingTime * 60))!
+            statusText = """
+            Distance remaining: \(distanceRemaining)
+            Time remaining: \(timeRemaining)
+            """
+            if status.currentManeuverIndex + 1 < directionsList.count {
+                let nextDirection = directionsList[status.currentManeuverIndex + 1].directionText
+                statusText.append("\nNext direction: \(nextDirection)")
+            }
+        case .reached:
+            mapView.locationDisplay.stop()
+            statusText = "Destination reached."
+            routeAheadGraphic.geometry = nil
+        default:
+            statusText = "Off route!"
+        }
+        updateRouteGraphics(remaining: status.routeProgress.remainingGeometry, traversed: status.routeProgress.traversedGeometry)
+        setStatus(message: statusText)
+    }
+    
+    func updateRouteGraphics(remaining: AGSGeometry?, traversed: AGSGeometry? = nil) {
+        routeAheadGraphic.geometry = remaining
+        routeTraveledGraphic.geometry = traversed
+    }
+}
+
+// MARK: - AGSLocationChangeHandlerDelegate
+
+extension NavigateRouteWithReroutingViewController: AGSLocationChangeHandlerDelegate {
+    func locationDataSource(_ locationDataSource: AGSLocationDataSource, locationDidChange location: AGSLocation) {
+        // Update the tracker location with the new location from the simulated data source.
+        routeTracker?.trackLocation(location)
+    }
+}
+
+extension NavigateRouteWithReroutingViewController: XMLParserDelegate {
+    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName: String?, attributes attributeDictionary: [String: String]) {
+        // Collect coordinates by checking for the lines that have a <trkpt> tag,
+        if elementName == "trkpt" {
+            //Create a World map coordinate from the file
+            let lat = Double(attributeDictionary["lat"]!)
+            let lon = Double(attributeDictionary["lon"]!)
+            let point = AGSPoint(x: lon!, y: lat!, spatialReference: .wgs84())
+            gpxPoints.append(point)
+        }
```

Comment:
```suggestion
        if elementName == "trkpt",
           let latString = attributeDictionary["lat"],
           let lonString = attributeDictionary["lon"],
           let lat = numberFormatter.number(from: latString)?.doubleValue,
           let lon = numberFormatter.number(from: lonString)?.doubleValue {
            // Create a World map coordinate from the file.
            let point = AGSPoint(x: lon, y: lat, spatialReference: .wgs84())
            gpxPoints.append(point)
        }
```
It is totally fine to force unwrap, as the gpx content is already known. But using conditional unwrapping makes it safer with little overhead.